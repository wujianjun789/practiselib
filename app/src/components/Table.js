import React, {Component} from 'react';

import '../../public/styles/table.less'
/**
 * Table 组件
 * keyField:string 指定key
 * columns:[{title,field},..others] 列属性显示，title：列名称，field：列属性
 * data:[{esn,name,...},..other]
 * allChecked:bool 全选或全取消属性
 * allCheckChange:Func(value:bool) 全选或全取消事件，value:true:全选，false:取消
 * rowCheckChange:Func(rowId:Num,value:bool) 选中或取消事件
 * rowEdit:Func(rowId:Num) 编辑事件
 * rowDelete:Func(rowId:Num) 删除事件
 */
export default class Table extends Component {
    constructor(props) {
        super(props);
        this.allCheckChange = this.allCheckChange.bind(this);
        this.rowCheckChange = this.rowCheckChange.bind(this);
        this.rowEdit = this.rowEdit.bind(this);
        this.rowDelete = this.rowDelete.bind(this);
        this.rowClick = this.rowClick.bind(this);
    }

    allCheckChange() {
        this.props.allCheckChange && this.props.allCheckChange();
    }

    rowCheckChange(rowId, value) {
        this.props.rowCheckChange && this.props.rowCheckChange(rowId, value);
    }

    rowEdit(rowId) {
        this.props.rowEdit && this.props.rowEdit(rowId);
    }

    rowDelete(rowId) {
        this.props.rowDelete && this.props.rowDelete(rowId);
    }

    rowClick(row){
        this.props.rowClick && this.props.rowClick(row);
    }

    render() {
        let {columns=[], data=[], allChecked, keyField='id', isEdit, className='', activeId} = this.props;
        return (
            <div className={`table-responsive ${className}`}>
                <table className="table table-hover">
                    <thead>
                    <tr>
                        <th className={allChecked === undefined?'hidden':''}><input type="checkbox"
                                                                                    checked={allChecked}
                                                                                    onChange={this.allCheckChange}/>
                        </th>
                        {
                            columns.map((item, index)=> {
                                return <th key={index}>{item.title}</th>
                            })
                        }
                        {
                            isEdit && <th></th>
                        }
                    </tr>
                    </thead>
                    <tbody>
                    {
                        data.map((row, index)=> {
                            let curId = row.get('id');
                            return <tr key={index} className={activeId && curId && activeId==curId ? 'active':''} onClick={()=>this.rowClick(row)}>
                                <td className={allChecked === undefined?'hidden':''}>
                                    {
                                        <input type="checkbox" checked={row.get('checked')} onChange={(e)=>keyField && this.rowCheckChange(row.get(keyField),e.target.checked)}/>
                                    }
                                </td>
                                {
                                    columns.map(function (item, index) {
                                        return <td key={index}>{row.get(item.field)}</td>
                                    })
                                }
                                {
                                    isEdit &&
                                        <td className="edit">
                                        <a className="btn" onClick={()=>keyField && this.rowEdit(row.get(keyField))}><svg><use xlinkHref={"#logo_update"} className="logo_update" transform="scale(0.075,0.075)" x="0" y="0" viewBox="1 -2 20 20" width="200" height="200"/></svg><span className="update">修改</span></a>
                                        <a className="btn" onClick={()=>keyField && this.rowDelete(row.get(keyField))}><svg><use xlinkHref={"#logo_del"} className="logo_del" transform="scale(0.08,0.08)" x="0" y="0" viewBox="1 -2 20 20" width="200" height="200"/></svg><span className="del">删除</span></a>
                                        </td>
                                }


                            </tr>
                        })
                    }
                    </tbody>
                </table>
            </div>
        )
    }
}