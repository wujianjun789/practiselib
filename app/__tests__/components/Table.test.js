/**
 * Created by a on 2017/7/10.
 */
import React from 'react';
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme';
import Table from '../../src/components/Table';

import Immutable from 'immutable'

const column=[{id:1, name:'name', field:'name'}, {id:1, address:'address', field:'address'}]
const data = Immutable.fromJS([{name:'xiaohong', address:'shanghai'}, {name:'xiaogang', address:'beijing'}])
test('table renders', ()=>{

    const component = renderer.create(
        <Table columns={column} data={data}/>
    )

    let table = component.toJSON();
    expect(table).toMatchSnapshot();
})

test('table tr click', ()=>{
    const click = jest.fn();
    const component = shallow(<Table columns={column} data={data} rowClick={click}/>)

    component.find('tr').at(1).simulate('click');
    expect(click).toHaveBeenCalledTimes(1);
})