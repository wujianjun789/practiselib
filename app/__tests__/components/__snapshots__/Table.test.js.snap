exports[`test table renders 1`] = `
<div
  className="table-responsive ">
  <table
    className="table table-hover">
    <thead>
      <tr>
        <th
          className="hidden">
          <input
            checked={undefined}
            onChange={[Function]}
            type="checkbox" />
        </th>
        <th />
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        onClick={[Function]}>
        <td
          className="hidden">
          <input
            checked={undefined}
            onChange={[Function]}
            type="checkbox" />
        </td>
        <td>
          xiaohong
        </td>
        <td>
          shanghai
        </td>
      </tr>
      <tr
        onClick={[Function]}>
        <td
          className="hidden">
          <input
            checked={undefined}
            onChange={[Function]}
            type="checkbox" />
        </td>
        <td>
          xiaogang
        </td>
        <td>
          beijing
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
